include/boost/geometry/algorithms/detail/overlay/get_turn_info_helpers.hpp

```diff
@@ -36,183 +37,309 @@ struct turn_operation_linear
     bool is_collinear; // valid only for Linear geometry
 };
 
-template <typename TurnPointCSTag, typename PointP, typename PointQ,
-          typename SideStrategy,
-          typename Pi = PointP, typename Pj = PointP, typename Pk = PointP,
-          typename Qi = PointQ, typename Qj = PointQ, typename Qk = PointQ
+template
+<
+    typename TurnPointCSTag,
+    typename UniqueSubRange1,
+    typename UniqueSubRange2,
+    typename SideStrategy
 >
 struct side_calculator
 {
-    inline side_calculator(Pi const& pi, Pj const& pj, Pk const& pk,
-                           Qi const& qi, Qj const& qj, Qk const& qk,
+    typedef typename UniqueSubRange1::point_type point1_type;
+    typedef typename UniqueSubRange2::point_type point2_type;
+
+    inline side_calculator(UniqueSubRange1 const& range_p,
+                           UniqueSubRange2 const& range_q,
                            SideStrategy const& side_strategy)
-        : m_pi(pi), m_pj(pj), m_pk(pk)
-        , m_qi(qi), m_qj(qj), m_qk(qk)
-        , m_side_strategy(side_strategy)
+        : m_side_strategy(side_strategy)
+        , m_range_p(range_p)
+        , m_range_q(range_q)
+    {}
+
+    inline int pk_wrt_p1() const { return m_side_strategy.apply(get_pi(), get_pj(), get_pk()); }
+    inline int pk_wrt_q1() const { return m_side_strategy.apply(get_qi(), get_qj(), get_pk()); }
+    inline int qk_wrt_p1() const { return m_side_strategy.apply(get_pi(), get_pj(), get_qk()); }
+    inline int qk_wrt_q1() const { return m_side_strategy.apply(get_qi(), get_qj(), get_qk()); }
+
+    inline int pk_wrt_q2() const { return m_side_strategy.apply(get_qj(), get_qk(), get_pk()); }
+    inline int qk_wrt_p2() const { return m_side_strategy.apply(get_pj(), get_pk(), get_qk()); }
+
+    inline point1_type const& get_pi() const { return m_range_p.at(0); }
+    inline point1_type const& get_pj() const { return m_range_p.at(1); }
+    inline point1_type const& get_pk() const { return m_range_p.at(2); }
+
+    inline point2_type const& get_qi() const { return m_range_q.at(0); }
+    inline point2_type const& get_qj() const { return m_range_q.at(1); }
+    inline point2_type const& get_qk() const { return m_range_q.at(2); }
+
+    SideStrategy m_side_strategy; // NOTE: cannot be const&
+    UniqueSubRange1 const& m_range_p;
+    UniqueSubRange2 const& m_range_q;
```

Comment:
So here `UniqueSubRange1` and `UniqueSubRange2` are `const&`, yet `SideStrategy ` is not. Are we sure that ranges will live as long as the side_calculator object? Are we sure that they will not be created from temporaries like:

    side_calculator sc(some_range1(), some_range2(), get_side_strategy());

because if they are created like that these references will be dangling.